slapo/schedule.py

```diff
@@ -899,27 +913,35 @@ def consolidate_model(sch: Schedule, topology=None):
             if is_found:
                 new_param = param.detach().split(sharded_size, dim=axis)[tp_rank]
                 sch.mod.register_parameter(param_name, nn.Parameter(new_param))
-    elif cnt_materialized != 0:
-        # skip if all the parameters have been materialized
-        assert (
-            cnt_meta == 0
-        ), f"Some of the parameters in module {sch.name} are on meta device"
-    elif cnt_meta == 0 and cnt_materialized == 0:
-        # skip if no direct parameters in this module
-        pass
-    for subsch in sch.child.values():
-        consolidate_model(subsch, topology)
+
+        for subsch in sch.child.values():
+            _consolidate_and_broadcast(subsch)
+
+    if cnt_meta != 0 or cnt_materialized != 0:
+        _consolidate_and_broadcast(sch)
+
+    gc.collect()
+    torch.cuda.empty_cache()
+
     return sch
 
 
-def build(sch: Schedule, topology=None, target=None, **kwargs):
+def build(
+    sch: Schedule,
+    topology=None,
+    target=None,
+    param_init_fn=None,
+    init_required: bool = True,
```

Comment:
`init_required` is a bit vague, and it is tightly coupled with `param_init_fn`. Can we name it like `init_weights: Optional[Union[bool, Callable]]`? It could be used as follows:

```python
# case 1: disable
init_weights=None or False
# case 2: enable with default init function
init_weights=True
# case 3: enable with custom init function
init_weights=model._init_weights
```

Reply:
I think it would be misleading for consolidation func. In consolidation func, param_init_fn should always be either None or an actual func, while init_required is used in build func to indicate if we would like to call consolidation func.

and using the example as you suggested, the init will be disabled by defaults. but I would expect users want initialization when they call build since build is mainly for the pipeline case. and if users are not familiar with the code, they may forget to enable that.

Reply:
`I would expect users want initialization when they call build since build is mainly for the pipeline case` that's not true. We expect users to call `build` in all cases. Although currently only pipeline needs it, we don't know if there's any requirements in the future. If you prefer to have this by default, we just need `init_weight: Union[bool, Callable]=True`, so that it becomes:

```python
# case 1 (default): enable with default init function
init_weights=True
# case 2: enable with custom init function
init_weights=model._init_weights
# case 3: disable
init_weights=False
```

Reply:
ok the 2nd point can be addressed by changing the default value, but the 1st point is still valid. consolidation by name means we want to initialize the model weights and synchronize them across the gpus when we call it. so making param_init_fn a True or False argument does not much sense.

Reply:
In my proposal there's no `param_init_fn` anymore, and `init_weights=[True, False, Callable]` seems not a problem to me?

Again my point is that users should call `.build` in all cases, so the `.build` API should also cover single device. Accordingly, it's not straightforward for users to understand the meaning of `param_init_fn` or `init_required` by their names. It is also easy for users to make mistake when the functionality of two arguments are tightly coupled. That's why I suggested 1) combining them to a single argument, and 2) having a better name. If `init_weights` isn't clear enough, we could come up with a better name, such as `consolidate_weights` or something like that.

Reply:
No, I am not talking about the name. What I mean is we will pass init_weights to consolidation_model func, and what is passed to consolidation_model shouldn't be a bool variable.

Reply:
Why not? `init_required` is also a bool variable.